components/places/src/import/ios_bookmarks.rs

```diff
@@ -0,0 +1,566 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::api::places_api::{PlacesApi, SyncConn};
+use crate::bookmark_sync::{
+    store::{BookmarksStore, Merger},
+    SyncedBookmarkKind,
+};
+use crate::error::*;
+use crate::types::SyncStatus;
+use rusqlite::{named_params, NO_PARAMS};
+use sql_support::ConnExt;
+use std::collections::HashMap;
+use url::Url;
+
+/// This import is used for iOS sync users migrating from `browser.db`-based
+/// bookmark storage to the new rust-places store.
+///
+/// It is only used for users who are not connected to sync, as syncing
+/// bookmarks will go through a more reliable, robust, and well-tested path, and
+/// will migrate things that are unavailable on iOS due to the unfortunate
+/// history of iOS bookmark sync (accurate last modified times, for example).
+///
+/// As a result, the goals of this import are as follows:
+///
+/// 1. Any locally created items must be persisted.
+///
+/// 2. Any items from remote machines that are visible to the user must be
+///    persisted. (Note: before writing this, most of us believed that iOS wiped
+///    it's view of remote bookmarks on sync sign-out. Apparently it does not,
+///    and it's unclear if it ever did).
+///
+/// Additionally, it's worth noting that we assume that the iOS tree is
+/// relatively well-formed. We do leverage `dogear` for the merge, to avoid
+/// anything absurd, but for the most part the validation is fairly loose. If we
+/// see anything we don't like (URL we don't allow, for example), we skip it.
+///
+/// ### Unsupported features
+///
+/// As such, the following things are explicitly not imported:
+///
+/// - Livemarks: We don't support them in our database anyway.
+/// - Tombstones: This shouldn't matter for non-sync users.
+/// - Queries: Not displayed or creatable in iOS UI, and only half-supported in
+///   this database.
+///
+/// Some of this (queries, really) is a little unfortunate, since it's
+/// theoretically possible for someone to care, but this should only happen for
+/// users:
+///
+/// - Who once used sync, but no longer do.
+/// - Who used this feature when they used sync.
+/// - Who no longer have access to any firefoxes from when they were sync users,
+///   other than this iOS device.
```

Comment:
(This is also a great explanation of why we import from the buffer, too, instead of just local!) 👍